design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,302 @@
+> +-------------------------+------------------------------------------------------------------------------------------------------+
+> | PARTITIONS              | The number of partitions in the backing topic. This property must be set if creating a SOURCE without|
+> |                         | an existing topic.                                                                                   |
+> +-------------------------+------------------------------------------------------------------------------------------------------+
+> | REPLICAS                | The number of replicas in the backing topic. This property must be set if creating a SOURCE without  |
+> |                         | an existing topic.                                                                                   |
+> +-------------------------+------------------------------------------------------------------------------------------------------+
+> ```
+* A new section for `INSERT INTO ... VALUES` will be added to the `sytnax.rst`:
+> ```rst
+> **Synopsis**
+> 
+> .. code:: sql
+>
+>   INSERT INTO <stream_name|table_name> [(column_name [, ...] )]
+>     VALUES (value [, ...])
+> 
+> **Description**
+> 
+> Produce a row into an existing stream or table and its underlying topic based on
+> explicitly sepcified values. The first ``column_name`` of every schema is ``ROWKEY``, which
+> defines the corresponding kafka key - it is expected to always match a corresponding column in 
+> the value if the create statement specifies a key.
+>
+> Column values can be given in several ways:
+> * Any column not explicitly given a value is set to ``null`` (KSQL does not support DEFAULT values)
+> * If no columns are specified, a value for every column is expected in the same order as the
+> schema
+>
+> The values will serialize using the ``value_format`` specified in the original `CREATE` statment.
+> ```
+* A new section for `DELETE` will be added to the `syntax.rst`:
+> ```rst
+> **Synopsis**
+> 
+> .. code:: sql
+> 
+>     DELETE FROM <table_name> WHERE ROWKEY = <value>; 
+>     
+> **Description**
+> 
+> Delete a row from an existing table. This will issue a tombstone, producing a value to the 
+> underlying kafka topic with the specified key and a null value. Deleting a value that 
+> does not exist will not affect the table contents.
+> ```
+
+
+# Compatibility Implications
+
+* All previous commands that succeeded will still succeed after this KLIP
+* `CREATE` commands that failed with "topic does not exist" may now fail with "must specify 
+partitions and replicas to create source with empty topic."
+* A new configuration `ksql.insert.into.values.enabled` will config guard this whole feature suite
+
+# Performance Implications
+
+N/A
+
+# Security Implications
+
+* Authentication/Authorization: we will authenticate/authorize `INSERT INTO` statements under the
```

Comment:
Can we be more explicit here?

Basically, to INSERT VALUES the user needs PRODUCE writes to the underlying topic. Simples.

Also, as mentioned above, this functionality is a prime candidate for being a RBAC operation so that different users may or may not have permissions to execute it.

Reply:
I wanted to avoid mentioning RBAC in open source documentation since it is a proprietary feature.

Reply:
Totally.